Add scheduling CLI

nilsvu opened the pull request:
## Proposed changes

Tool for scheduling runs, submitting jobs, setting up directories for segments and checkpoints, restarting from checkpoints, resubmitting jobs. Also allows to schedule ranges of runs, e.g. for convergence tests.

Try it like this:

```sh
spectre schedule INPUT_FILE -o ./Run
```

To run a convergence test put something like `InitialGridPoints: {{ num_points }}` in your input file and try this:

```sh
spectre schedule INPUT_FILE -p num_points=6...8 -o './Run_p{{ num_points - 1 }}'
```

Closes #2987.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
- Install `Jinja2` in your Python environment (it's probably already installed because other packages need it).
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add Jinja2 to Py requirements

To work with placeholders in input files and submit scripts.

nilsvu commented:
Notes from a discussion with @wthrowe @knelli2: We decided to look at template inheritance https://jinja.palletsprojects.com/en/3.1.x/templates/#template-inheritance to solve the issue with generating submit scripts.

wthrowe commented:
Following up on discussion elsewhere: The restart option can be added without duplicating the rest of the command with `${SPECTRE_CHECKPOINT:+ +restart "${SPECTRE_CHECKPOINT}"}`.  (The `:+` expansion gives the stuff to its right, but only if the variable on the left is defined and nonempty.)

nilsvu commented:
I implemented the submit script generation, take a look when you get the chance

knelli2 commented on src/Parallel/Main.hpp:
```diff
@@ -140,15 +140,15 @@ class Main : public CBase_Main<Metavariables> {
+  // Return the dir name for the Charm++ checkpoints as well as the prefix for
+  // checkpoint names and their padding. This is a "detail" function so that
+  // these pieces can be defined in one place only.
+  std::tuple<std::string, std::string, size_t> checkpoints_dir_prefix_pad()
```
Once you make this a full fledged PR, can you factor these changes into another commit? This is just changing out checkpoint directory structure.

knelli2 commented on support/SubmitScripts/SubmitBase.sh:
```diff
@@ -0,0 +1,79 @@
+echo "Executable: ${SPECTRE_EXECUTABLE}"
+echo "CLI: ${SPECTRE_CLI}"
+echo
+
+/usr/bin/modulecmd bash list
```
I think this is specific to Wheeler.

nilsvu replied:
Any reason why this isn't `module list`?

knelli2 replied:
Because for some reason, `module list` didn't work on the unlimited queue on Wheeler

knelli2 commented on support/SubmitScripts/SubmitBase.sh:
```diff
@@ -0,0 +1,79 @@
+    --submit-script-template "{{ submit_script_template }}" \
+    --executable ${SPECTRE_EXECUTABLE} \
+    --num-nodes ${SLURM_NNODES} \
+    --queue ${SLURM_JOB_PARTITION} \
+    --time-limit $(squeue -j ${SLURM_JOB_ID} -h --Format TimeLimit) \
```
Is there a slurm env variable that will give you the current time limit?

nilsvu replied:
Unfortunately no :(

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
```
Just a general comment: Would we be able to break up this file either into smaller files for the various classes/functions and maybe even separate commits? This is a lot for one commit and it's some dense code to go through.

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+            executable = metadata["Executable"]
+        except KeyError as err:
+            raise ValueError("Specify an 'executable' ('--executable' / '-e') "
+                             "or list one in the input file.") from err
```
`"in the input file metadata as 'Executable:'"`

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+            run_dir = Path.cwd()
+        else:
+            raise ValueError("Specify a 'run_dir' ('--run-dir' / '-o') "
+                             "or a 'segments_dir' ('--segments-dir' / '-O').")
```
`"or place an input file into the current directory if you aren't using a scheduler"`.

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+    if segments_dir:
+        context.update(segments_dir=segments_dir.resolve())
+
+    # Resolve outfile
+    out_file = run_dir / out_file_name
```
What is this syntax? Is it specific to `Path`?

nilsvu replied:
Yep, https://docs.python.org/3/library/pathlib.html

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+    if not (input_file_path.resolve() == input_file_template.resolve()
+            and rendered_input_file == input_file_contents):
+        if input_file_path.exists() and not force:
+            raise OSError(
+                f"File already exists at '{input_file_path}'. Delete it, "
```
`"Input file already exists"`?

nilsvu replied:
Moved to generic function

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+        original_executable = executable
+        executable = (segments_dir / executable.name).resolve()
+        if not force and executable.exists():
+            raise OSError(
+                f"File already exists at '{executable}'. Delete it, "
```
`"Executable already exists"`?

nilsvu replied:
Moved to generic function

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+    submit_script_template_env = template_env.overlay(
+        loader=jinja2.FileSystemLoader(submit_script_template.parent))
+    # Copy submit script templates to segments directory
+    if segments_dir and submit_script_template.parent != segments_dir.resolve(
+    ):
```
Is this really what yapf decided was the best formatting???

nilsvu replied:
Yep :D Gotta switch to black at some point..

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+        submit_script_template = (segments_dir /
+                                  submit_script_template.name).resolve()
+        if not force and submit_script_template.exists():
+            raise OSError(
+                f"File already exists at '{submit_script_template}'. "
```
`"Submit script template"`?

nilsvu replied:
Moved to generic function

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+        shutil.copy(original_submit_script_template, submit_script_template)
+        submit_script_template_env = template_env.overlay(
+            loader=jinja2.FileSystemLoader(segments_dir))
+        ast = submit_script_template_env.parse(
+            submit_script_template.read_text())
```
What's `ast`?

nilsvu replied:
"abstract syntax tree" https://jinja.palletsprojects.com/en/3.1.x/api/#the-meta-api

knelli2 replied:
maybe name this `syntax_tree` then? Just to at least have some descriptor of what it is (even if I don't know what it actually is)

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+            referenced_template_dst = segments_dir / referenced_template
+            if not force and referenced_template_dst.exists():
+                raise OSError(
+                    f"File already exists at '{referenced_template_dst}'. "
```
`"Submit script template"`? (I don't know how this one differs from the previous one)

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+            and rendered_submit_script == submit_script_template.read_text()):
+        if not force and submit_script_path.exists():
+            raise OSError(
+                f"File already exists at '{submit_script_path}'. Delete it, "
```
`"Submit script"`

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+                                        cwd=run_dir,
+                                        capture_output=True,
+                                        text=True)
+        matched_submit_msg = re.match(r"Submitted batch job (\d+)",
+                                      submit_process.stdout)
```
Is this the guaranteed message on all clusters?

nilsvu replied:
Haven't seen Slurm output anything else, but we can change this if we run into issues

wthrowe replied:
Does it give a useful return status?  That should be more robust than text parsing.

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+
+    - Integers or floats
+    - List of values: "[1, 2, 3]"
+    - Exclusive range: "0..3" or "0..<3" (the latter is clearer, but "<" is a
+      special character in the shell)
+    - Inclusive range: "0...3"
```
Is this standard notation? In bash `{0..5}` is inclusive.

nilsvu replied:
I got this from [Swift](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/basicoperators#Range-Operators) where `1..<3` is a half-open (exclusive) range and `1...3` is a closed (inclusive) range. Ranges are often a source of confusion and ambiguity, and I think the Swift syntax is quite clear.

knelli2 replied:
If this follows Swift syntax, maybe add links to Swift in the relevant places so people can go look at that as well.

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+import yaml
+
+logger = logging.getLogger(__name__)
+
+default_submit_script_template = Path(__file__).parent / 'Submit.sh'
```
Where is this? I don't know what the `.parent` method of a Path does, but this looks like it's looking for a file named `Submit.sh` in the `support/` directory? Oh also, wouldn't this behavior change if you're in PY_DEV_MODE or not because of linking?

nilsvu replied:
Yep it's next to the `Schedule.py` file in `support/`. That's were CMake puts the machine-specific submit script template. It works with `PY_DEV_MODE` because the `.parent` is purely lexical (doesn't resolve symlinks). I'm open to better ideas how to copy around submit scripts!

knelli2 replied:
Ok this seems fine then

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+    help=("Path to a submit script. "
+          "It will be copied to the 'run_dir'. It can be a [Jinja template]("
+          "https://jinja.palletsprojects.com/en/3.0.x/templates/) "
+          "(see main help text for possible placeholders)."))
+@click.option('--queue', help=("Name of the queue."))
```
I forget, is the plan to get these defaults (queue, time limit, nodes/cores) from the Machine file? Or are we getting rid of that?

nilsvu replied:
At the moment I got rid of any use of the machine files in this script. Instead, there's only the submit script template (one per cluster), which sets these defaults. As it stands now we could delete the machine files.

knelli2 replied:
Ok I'll test it out and see how it works

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,779 @@
+              show_default="executable name",
+              help=("A short name for the job "
+                    "(see main help text for possible placeholders)."))
+@click.option('--num-procs',
+              '-j',
```
Maybe `-c`? I know this is make syntax, but `-c` is srun syntax. Not sure which is better.

nilsvu replied:
I don't know either. I don't think there's a consensus among other tools. I got used to `-j` because of make. We could allow both, if that helps.

knelli2 replied:
Yeah do both

knelli2 reviewed (changes requested):
Haven't tested this out yet, but did a first pass of review. I'll probably have more things as I test it out.

nilsvu commented:
Thanks @knelli2, I implemented your suggestions and squashed because I moved commits / files / functions around. Sorry that that makes the smaller changes harder to find!

nilsvu committed:
Generate Wheeler and CaltechHPC submit scripts

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,702 @@
+
+    1. 'job_name' (if specified):
+
+        - All arguments to this function, including all additional '**kwargs'.
```
These indents are rendered weird.

```
- All arguments to this function, including all additional '--params'.
      For example, the additional '--params' can include parameters
      controlling resolution in the input file.     - 'executable_name': Just
      the name of the executable (basename of the       'executable' path).
```

Same for the rest below.

wthrowe commented on src/Parallel/Main.hpp:
```diff
@@ -941,39 +942,37 @@ void Main<Metavariables>::post_deadlock_analysis_termination() {
+Main<Metavariables>::checkpoints_dir_prefix_pad() const {
+  const std::string checkpoints_dir = "Checkpoints";
+  const std::string prefix = "Checkpoint";
+  constexpr size_t pad = 3;
```
Leave it at 6.  Characters are cheap.

nilsvu replied:
Yea they're cheap but become hard to read/count at some point. Try this: just glancing at `0000` you immediately know that it's 4 zeros. But glancing at `00000` you don't, you have to actually count. So I can make it 4 characters if you think we need that many, but anything beyond that I'd like to avoid. (also keep in mind that the checkpoint/segment names are actually not set in stone, I want to make sure that none of the code really relies on a hard-coded directory structure but just takes as input the files that it needs).

wthrowe replied:
OK, but why do you want to count the digits?  I don't see a reason that should be a design goal.

If you don't want to discuss the counter format in this PR then don't change it.

nilsvu replied:
I think readable directory names are a design goal. No need for excessive digits in my opinion. More important for segments than for checkpoints, so if you prefer I can keep the checkpoint format unchanged in this PR.

wthrowe replied:
I hadn't actually compared them, so I tried creating a list of `CheckpointNNN` and `CheckpointNNNNNN` names from 0 to 19 (I got tired of typing at that point).  I actually found the six-digit version to be significantly easier to read, because the string of zeros acts as a separator between "Checkpoint" and the number.  It makes it easier to visually locate the first nonzero digit.  Presumably the same would hold for other prefixes.

[Continuing from a thread below, but consolidating because the topic is the same]

> Three digits give us over 3 years in daily segments. That seems enough for the vast majority of simulations. As mentioned above the number of digits is not set in stone, so if we want to do a 5-year simulation or do hourly segments for some reason or something like that we can always increase the number of digits for that case.

You assume that segments will always be based on a timer.  This has not always been true for SpEC, and while it is a goal for SpECTRE I don't think we should assume that we will always achieve it immediately.  SpEC segments could be extremely short (minutes, maybe less) because they were based on some event in the simulation that forced a new segment.

Anyway, @markscheel, you're probably the most experienced SpEC-runner here.  Do you want to weigh in?

nilsvu replied:
I see what you mean with the zero-padding visually separating the number from the prefix. How about 4 digits for both segments and checkpoints then? Still not excessive, hence ok to type without having to pay too much attention to the number of keystrokes (which I find kind of hard for 6 digits), but surely enough and visually pleasing.

wthrowe commented on support/Python/DirectoryStructure.py:
````diff
@@ -0,0 +1,116 @@
+    ```
+
+    With SpEC we never needed more than 26*26=676 segments (AA-ZZ), so three
+    digits for segments is enough.
````
Is this true?  I vaguely think I've seen three-character SpEC segment IDs.

nilsvu replied:
I haven't seen those, but that doesn't say much.

Three digits give us over 3 years in daily segments. That seems enough for the vast majority of simulations. As mentioned above the number of digits is not set in stone, so if we want to do a 5-year simulation or do hourly segments for some reason or something like that we can always increase the number of digits for that case.

wthrowe replied:
I've addressed the design comment in the other thread to try to keep the discussion more organized.

The point was trying to make here wasn't about your proposed design, but that I don't think "With SpEC we never needed more than 26*26=676 segments" is true, so you shouldn't claim it no matter what we decide for SpECTRE.

wthrowe commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,704 @@
+        if not force:
+            raise OSError(f"File already exists at '{path}'. Delete it or "
+                          "retry with 'force' ('--force' / '-f')." +
+                          (("\n" + error_hint) if error_hint else ""))
```
[optional] If we were confident that deleting the file was the correct thing to do, we would just do it here, so telling the user to do so may be giving bad instructions.  I'd just leave it at "File already exists.  Retry with --force to overwrite.".

wthrowe commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,704 @@
+    dest = (dest_dir / src_file.name).resolve()
+    if dest.exists() and not force:
+        raise OSError(f"File already exists at '{dest}'. Delete it or "
+                      "retry with 'force' ('--force' / '-f').")
```
Same comment as above.

wthrowe reviewed (changes requested):
Why does wheeler have special instructions in the initial PR comment?  Is that still relevant, or have things become more uniform?

wthrowe commented:
Forgot to give my overall comment: I'm very happy with how the submit scripts are coming out with the templates.  This fixes the various problems with inconsistencies and duplication that we currently have, so this resolves the initial issue in #2987, although that spawned a lot of extra discussion that may not be considered resolved.

nilsvu commented:
Pushed a fixup @wthrowe

knelli2 commented on src/Visualization/Python/ReadInputFile.py:
```diff
@@ -44,3 +41,25 @@ def find_event(event_name: str, input_file: dict) -> dict:
+    if not "PhaseChangeAndTriggers" in input_file:
+        return None
+    for trigger_and_phase_changes in input_file["PhaseChangeAndTriggers"]:
+        for phase_change in trigger_and_phase_changes["PhaseChanges"]:
+            if isinstance(phase_change, str):
```
The phase change `CheckpointAndExitAfterWallclock` is a string though and we want that returned if it's there

nilsvu replied:
Oh I just used it wrong, the empty dict that's returned here evaluates to False in an if condition, so I changed it to `is not None`

knelli2 replied:
Ok now it works!

knelli2 reviewed (changes requested):
Couple things. You can ignore the formatting one for now, but the phase change comment is because it's not behaving correctly. I have `CheckpointAndExitAfterWallclock` in my input file when trying to schedule just a stand alone run dir, and it doesn't error.

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -411,13 +420,43 @@ def schedule(input_file_template: Union[str, Path],
+        "CheckpointAndExitAfterWallclock", input_file)
+    if wallclock_exit_phase_change is not None and not segments_dir:
+        raise ValueError(
+            "Found 'CheckpointAndExitAfterWallclock' in the input file but "
+            "no 'segments_dir' ('--segments-dir' / '-o') is set. "
```
`-o` is for run dir. `-O` is for segments dir

nilsvu replied:
oops

wthrowe commented on support/Python/Schedule.py:
```diff
@@ -226,6 +230,15 @@ def schedule(input_file_template: Union[str, Path],
+        the run directory if no segments directory is set (it still won't be
+        copied to individual segments in a segments directory). This behavior
+        supports reproducibility.
+        If 'copy_executable' is set to False, the executable is never copied.
```
I'm confused.  Is the default true or false?  Is this saying the default depends on whether there is a segments directory?  (Which would be fine, but it should make that clear.)  Similarly confused by the short description below.

nilsvu replied:
Pushed another fixup to clarify

wthrowe commented on docs/Tutorials/CheckpointRestart.md:
````diff
@@ -31,9 +31,9 @@ Executables can checkpoint when:
+./MySpectreExecutable +restart Checkpoints/Checkpoint123
 ```
-where the `000123` should be the number of the checkpoint to restart from.
+where the `123` should be the number of the checkpoint to restart from.
````
Whatever we decide in the discussion on formatting, this documentation should make it clear that these are zero-padded, either by stating it or by using an example with a leading zero.

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,754 @@
+          "unless starting the first segment."))
+@click.option('--copy-executable/--no-copy-executable',
+              default=None,
+              help=("Copy the executable to the run or segments directory. "
+                    "Default (when no flag is specified) is to copy only to "
+                    "the segments directory to support resubmission. "
+                    "With '--copy-executable' the executable is also copied "
+                    "to the run directory (only has an effect if "
+                    "'--run-dir' / '-o' is set). "
+                    "With '--no-copy-executable' the executable is never "
+                    "copied."))
```
This is still a bit confusing. May I suggest

```
Copy the executable to the run or segments directory.
When no flag is specified:
 - If '--run-dir' / '-o' is set, don't copy.
 - If '-segments-dir' / '-O' is set, copy to segments directory to
   support resubmission.
If '--copy-executable' is specified:
 - If '--run-dir' / '-o' is set, copy to the run directory.
 - If '-segments-dir' / '-O' is set, copy to segments directory to
   support resubmission. (Note: the executable is not copied to
   individual segments)
If '--no-copy-executable' is specified: Never copy.
```

I know it's a bit verbose, but I think this is very clear on what the behavior is and is easy for users to follow.

nilsvu replied:
Sounds good, done.

nilsvu commented:
@wthrowe @knelli2 lots of fixups for you to look at whenever you get the chance :)

knelli2 commented on support/Python/CMakeLists.txt:
```diff
@@ -25,7 +25,7 @@ if(MACHINE)
     )
   configure_file(
     ${CMAKE_SOURCE_DIR}/support/SubmitScripts/${MACHINE}.sh
-    ${SPECTRE_PYTHON_PREFIX}/support/Submit.sh
+    ${SPECTRE_PYTHON_PREFIX}/support/SubmitTemplate.sh
```
Also need to change this towards the top of `Status.py`:
```py
default_submit_script_template = Path(__file__).parent / "SubmitTemplate.sh"
```
and a couple places in comments/docs I think.

I also still think `SubmitBase.sh` should be `SubmitTemplateBase.sh`. Even though now it's clear neither should be used as a submission script, I have no clue what the difference between them is. I think having `NameBase.sh` and `Name.sh` is clear that `Name.sh` depends on `NameBase.sh` (and then the `Template` part still signals that it shouldn't actually be used to submit).

nilsvu replied:
ok

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -349,7 +351,10 @@ def schedule(
 
     # Resolve CLI
-    spectre_cli = shutil.which("spectre")
+    # This is the path of the `spectre` CLI where this script is installed.
+    # We may have to make this more robust for resubmissions if we run into
+    # problems or unexpected behavior.
+    spectre_cli = Path(__file__).parent.parent.parent.parent / "spectre"
```
This seems very fragile (also build directories don't have to be placed inside the SPECTRE_HOME). What was wrong with the previous method?

nilsvu replied:
This isn't as fragile as you think because this is in the build directory (not in `SPECTRE_HOME`). So it only breaks if we move the location of the python package inside the build dir.

This method ensures that we use the same CLI for resubmissions. The previous method would use whichever CLI is in the `PATH`, which we can't control.

nilsvu replied:
Added a test for this too

knelli2 replied:
This will work even in `PY_DEV_MODE`?

nilsvu replied:
Yep, symlinks aren't resolved

knelli2 commented on support/SubmitScripts/SubmitBase.sh:
```diff
@@ -0,0 +1,81 @@
+    --executable ${SPECTRE_EXECUTABLE} \
+    --num-nodes ${SLURM_NNODES} \
+    --queue ${SLURM_JOB_PARTITION} \
+    --time-limit $(squeue -j ${SLURM_JOB_ID} -h --Format TimeLimit) \
+    --from-last-checkpoint ./Checkpoints \
```
This must be

```sh
    --from-last-checkpoint ${RUN_DIR} \
```

The current way fails.

nilsvu replied:
Worked for me, did this fail when you tried? The code should resolve the relative path correctly (we're in the `RUN_DIR` here).

knelli2 replied:
It had previously failed, but now it works. Not sure what changed.

knelli2 commented on support/Python/Schedule.py:
```diff
@@ -415,16 +415,21 @@ def schedule(
+                " checkpoint. This is technically possible, but probably"
+                " wrong. This assert safeguards against inconsistent usage."
+                " Either run in a different directory, or use"
+                " '--from-checkpoint' to continue from the last checkpoint."
```
Hmm this error message is trying to do two things at once and it makes things confusing. There are two ways that I run into this:

1. I'm trying to continue from a checkpoint that is not the last checkpoint in a `Checkpoints` directory.
2. I try to continue from a segment that is not the most recent segment in a segments directory.

I still think we shouldn't allow either at the moment, but we need to distinguish between the two. What about this:

```py
        assert from_checkpoint, "No checkpoint specified, but trying to restart from a checkpoint"
        last_segment = all_segments[-1]
        if Path(from_checkpoint).parent.parent.resolve() != last_segment.path.resolve():
            raise AssertionError(
                "You're not continuing from the latest segment "
                f"{last_segment.path.resolve()}. All segments:\n"
                f"{[seg.path.resolve() for seg in all_segments]}"
            )
        last_checkpoint = all_segments[-1].checkpoints[-1]
        if Path(from_checkpoint).resolve() != last_checkpoint.path.resolve():
            raise AssertionError(
                "You're not continuing from the latest checkpoint "
                f"{last_checkpoint.path.resolve()}. All checkpoints:\n"
                f"{[check.path for check in all_segments[-1].checkpoints]}"
            )
```

Feel free to make formatting nicer/more verbose messages, but I'd say keep the two messages separate.

nilsvu replied:
I split this error message into three, but kept the "last segment" vs "last checkpoint" together. The error prints the expected path of the checkpoint, which includes both the segment and the checkpoint, so I see no point in splitting it into separate errors messages.

knelli2 replied:
I still think the third error message you have isn't clear.

```
You're not continuing from the previous segment's last checkpoint
(Segment001/Checkpoints/Checkpoint000).
```

I agree that this is a true statement and that it tells you what you should be doing (i.e. continuing from this specific checkpoint which is the last checkpoint in the latest segment). However, it's confusing because it doesn't tell me what I did wrong. Did I specify a segment other than the latest one? Or did I specify the correct segment, but not the last checkpoint in the latest segment?

And to me, that's confusing because I'm going to go back to the (incorrect) command I used and look for the offending option. If I specified an incorrect segment, that's what I'll be looking for. If it's an incorrect checkpoint, then I'll be looking for that. But I have to know what I'm looking for in the first place and your error message doesn't tell me that.

Also, I didn't find this line very helpful.

```
Either run in a different directory, or use '--from[-last]-checkpoint'
to continue from the last checkpoint.
```

Again, what I should be doing is dependent on whether I specified the incorrect segment or checkpoint. If I specified the correct segment but an incorrect checkpoint, the suggestion is simple:

```
Use '--from-last-checkpoint {last_segment}' or '--from-checkpoint {last_checkpoint}' to
continue from the last checkpoint.
```

However, if I specified an incorrect segment it's a bit trickier because of potentially corrupted segments. Did I actually want to continue from the segment I specified? If so, then I need to remove all segments after that one. Or did I just accidentally specify the wrong segment and I meant to continue from the latest one? For both of these I think a message like this would be helpful:

```
You are not continuing from the latest segment {last_segment}. If you want to continue
from the last checkpoint in the segment you specified ({specified_segment}), you must
remove all the segments after that one [list segments to remove?]. Otherwise, specify
'--from-last-checkpoint {last_segment}' to continue from the latest segment.
```

So overall I think this should be the order of the `asserts`:

1. assert from_checkpoint, "Found segments. Use '--from-last-segment {last_segment}' to continue from the last checkpoint in {last_segment}"
2. assert we are in last segment. If not, use above message.
3. assert current segment (which is the latest segment) has checkpoints, "Oh no, no checkpoints to continue from"
4. assert continuing from last checkpoint in this segment (the latest segment). if not use above message.

I know this is a lot of logic and similar error messages whereas yours is less logic and combined messages. But I think the differences in the message that are tailored to the specific errors are very clear, will help users a lot, and make the experience of using the scheduler a lot smoother. What do you think?

nilsvu replied:
See fixup

knelli2 replied:
Sorry, but I still think these error messages will be confusing for users. However, I've held up this PR long enough and the messages are already decent-ish so I think it's better to just get this merged and I can do a follow-up PR.

knelli2 reviewed (changes requested):
Couple more comments about some of the current fixups.

nilsvu commented:
@knelli2 please check the latest fixups when you get the chance, then I can squash them all together for @wthrowe to look at (so he doesn't have to read through a million fixups)

nilsvu commented:
Pushed another fixup for resubmissions ✌️

knelli2 commented:
Ok you can squash now. I imagine there will be bugs we'll need to iron out, but should be ok for now

nilsvu commented:
Squashed it all down. @wthrowe there were so many fixups that I lost track of what you've seen and what you haven't. I can probably dig up the git reflog if you prefer to see a single squashed-together fixup on top of your review above.

wthrowe commented:
The test is failing.

wthrowe commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,1064 @@
+    incrementing integers and continue the run from the previous segment. For
+    example, the following is a typical 'segments_dir':
+
+    \b
```
This is probably some python thing that I don't understand, but everything I can find on the internet says this is adding a backspace character.  Why?

nilsvu replied:
The `\b` escape marker prevents [click](https://click.palletsprojects.com/en/8.1.x/documentation/#preventing-rewrapping) from rewrapping the following paragraph: https://click.palletsprojects.com/en/8.1.x/documentation/#preventing-rewrapping

wthrowe commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,1064 @@
+      copy_executable: Copy the executable to the run or segments directory.
+        By default (when set to 'None'):
+          - If '--run-dir' / '-o' is set, don't copy.
+          - If '-segments-dir' / '-O' is set, copy to segments directory to
```
Typo in number of dashes.  Again 4 lines down.

wthrowe commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,1064 @@
+
+    Returns: The 'subprocess.CompletedProcess' representing either the process
+      that scheduled the run, or the process that ran the executable if
+      'scheduler' is 'None'. Returns 'None' if no run was scheduled.
```
What if multiple runs are scheduled?

nilsvu replied:
Changed to `Returns 'None' if no or multiple runs were scheduled.`

nilsvu commented:
Rebased!

wthrowe commented on support/Python/Schedule.py:
```diff
@@ -0,0 +1,1064 @@
+        "Copy the executable to the run or segments directory. "
+        "(1) When no flag is specified: "
+        "If '--run-dir' / '-o' is set, don't copy. "
+        "If '-segments-dir' / '-O' is set, copy to segments "
```
Dashes again.

wthrowe reviewed (changes requested):
Mostly looked at the documentation, since I'm not good with python.

nilsvu committed:
Add scheduling script to CLI

nilsvu committed:
Status CLI: get input file from submit context

nilsvu committed:
Add resubmit command to CLI

nilsvu commented:
Squashed in those fixes and split the changes into a few more commits

knelli2 reviewed (approved):
Looks good! Excited to start using this

nilsvu commented:
@wthrowe I think this is ready to go
